arroyo/processing/strategies/filter.py

```diff
@@ -1,35 +1,105 @@
+    payload, but only carry forward partition offsets for later strategies to
+    commit. Those messages have a payload of type `FilteredPayload`.
+
+    For that reason, basically every strategy needs to be able to handle
+    `Message[Union[FilteredPayload, T]]` instead of `Message[T]`, i.e. it needs
+    to subtype `ProcessingStrategy[Union[FilteredPayload, TStrategyPayload]]`.
+    If it doesn't, and rather just handles the regular `Message[T]`, it cannot
+    be composed with this step, and many other default strategies of arroyo.
+
+    If no `CommitPolicy` is passed, no "sentinel messages" are emitted and
+    downstream steps do not have to deal with such messages (despite the type
+    system telling them so).
     """
 
     def __init__(
         self,
-        function: Callable[[Message[TPayload]], bool],
-        next_step: ProcessingStrategy[TPayload],
+        function: Callable[[Message[TStrategyPayload]], bool],
+        next_step: ProcessingStrategy[Union[FilteredPayload, TStrategyPayload]],
+        commit_policy: Optional[CommitPolicy] = None,
     ):
         self.__test_function = function
         self.__next_step = next_step
 
+        if commit_policy is not None:
+            self.__commit_policy_state: Optional[
+                CommitPolicyState
+            ] = commit_policy.get_state_machine()
+        else:
+            self.__commit_policy_state = None
+
+        self.__uncommitted_offsets: MutableMapping[Partition, int] = {}
         self.__closed = False
 
     def poll(self) -> None:
         self.__next_step.poll()
 
-    def submit(self, message: Message[TPayload]) -> None:
+    def submit(
+        self, message: Message[Union[FilteredPayload, TStrategyPayload]]
+    ) -> None:
         assert not self.__closed
 
-        if self.__test_function(message):
+        now = time.time()
+
+        if not isinstance(message.payload, FilteredPayload) and self.__test_function(
+            cast(Message[TStrategyPayload], message)
+        ):
+            for partition in message.committable:
+                self.__uncommitted_offsets.pop(partition, None)
             self.__next_step.submit(message)
+        elif self.__commit_policy_state is not None:
+            self.__uncommitted_offsets.update(message.committable)
+
+        policy = self.__commit_policy_state
+
+        if policy is None or policy.should_commit(now, message.committable):
+            self.__flush_uncommitted_offsets(now)
+
+    def __flush_uncommitted_offsets(self, now: float) -> None:
+        if not self.__uncommitted_offsets:
+            return
+
+        new_message: Message[Union[FilteredPayload, TStrategyPayload]] = Message(
+            Value(FILTERED_PAYLOAD, self.__uncommitted_offsets)
+        )
+        self.__next_step.submit(new_message)
+
+        if self.__commit_policy_state is not None:
+            self.__commit_policy_state.did_commit(now, self.__uncommitted_offsets)
+
```

Comment:
Every message would still be flushed individually if no commit policy is passed. But the rationale is to cap the number of filtered messages that potentially have to be processed by every other subsequent strategy by just batching them up.

Reply:
Thinking about it again though, it may actually be a lot simpler to forget about all that and just forward all the messages. Then there'd be no need to provide a commit policy here, it'd get handled later anyway.

Reply:
This entire problem with filtered messages causing consumer lag only exists in specific scenarios where one filters out a lot of messages. Outside of that IMO fringe usecase, forwarding filtered messages at all is just additional load on subsequent strategies that serves no purpose.

If one doesn't pass a commit policy, the intent was that there are no FILTERED_PAYLOADs being produced. This default behavior minimizes risk of breakage in consumers that use FilterStep and have weak type checking.

Reply:
seems like this was actually not implemented as intended, fixed